### PR TITLE
Gangs Rework - Gamemode

### PR DESCRIPTION
**Pull request by TDHooligan**
## About the PR 

Gangs rework, if you want more depth, there's a [big ol' post here with ALL changes](https://discord.com/channels/182249960895545344/1024758522518446211/1024758522518446211), and the rationales behind them, if that's your thing. 

New Items:
 - Quickhack (**'ai open'**s up to 5 doors)
 - Janktank II (Revives a dead gang member, albeit in a dodgy state)

New Mechanics:
 - Tagging
 - Laundering 
 - Crate/Loot Drops
 - Leader Dropouts (cryo)

Smaller features:
 - Pop-ups for gang leaders
 - Ghosts can view gang territories on gang rounds
 - Gang leaders have a different icon to gang members
 - Gangs have uniquely colored antag icons

## Why's this needed? 

woe betide ye security mains. the greytide is coming once more

## Changelog 

```changelog
(u)TDHooligan
(*) Gangs has been reworked! Read below for deets. Send feedback in the discord thread or forum. 
(*) Gang member point pools are separate now. Huzzah!
(*) Gang lockers can no longer be broken.
(*) Gang leaders can buy the new 'Janktank II' to revive fellow gang members
(*) You can now launder thousands of credits at once. 
(*) Tagging has been reworked, read below for details.
(*) New gang loot bags & crates are hidden throughout the shift. Watch your radios, gang!
(+) Territory is now a circle around tags, and tags now provide points based on how many people they see pass by.
(+) You can only tag near your existing territory, starting where your locker is placed
(+) You should look to expand to populated areas, as quickly as possible!
(+) Your uniform also grants buffs if you're in your territory, and regen if you're by your locker.
(+) ---
(+) Every point earned also grants the gang 1 point of 'Street Cred'
(+) Street cred can be spent by leaders to hire new gang members, or buy revival syringes for their gang.
(+) ---
(+) Check your gang radio for new events:
(+) A few times a shift, loot bags are randomly hidden. Interrogate civilians, or their PDAs, to find where they are!
(+) Twice per shift, a large weapons cache will spawn, locked in place. fight to drag it back to your locker!
(+) ---
(+) Sprites by Mora, Guv, JOEL, Blackrep, what, Moriarty, Rubi, Shmiggy and Cogwerks. thank you so much :)
```

### CODE REVIEW

**Comment by Flaborized** [comment]
Hi haven't followed the development on this particularly closely but read through the big writeup; the feature ideas here are awesome and seem very fun, great work. Gang definitely needs the improvements and this seems like a really good direction to pull it in. I didn't read the code though so this is not a code review :shelterfrog:

**Comment by TobleroneSwordfish** [comment]
Some initial jank in no particular order:
- Pill bottles should be able to go in gang lockers.
- The message for pouring a beaker of drugs into a gang locker claims you put the beaker into the locker.
- The gang locker is non-dense. I think we talked about this at some point, but discord won't let me search threads properly so uhh why is it non-dense?
- The round seems to end instantly when the shuttle is auto-called?
- The little points popup doesn't display when putting drugs into a locker.

File: code/datums/gamemodes/gangwar.dm [hunk 1]
```diff
@@ -637,31 +952,74 @@ proc/broadcast_to_all_gangs(var/message)
...
-		var/turf/turftarget = get_turf(target)
-
-		if((turftarget == loc) || (BOUNDS_DIST(src, target) > 0))
-			return
+	var/in_use = FALSE
+	var/empty = FALSE
+
+	/// Checks a tile has no nearby claims from other tags
+	proc/check_tile_unclaimed(turf/target, mob/user)
+		for (var/obj/decal/gangtag/tag in range(GANG_TAG_SIGHT_RANGE,target))
```
Comment (TobleroneSwordfish) [inline code/datums/gamemodes/gangwar.dm:960]: Doesn't matter tooo much but it'd almost certainly be faster to iterate all gang tags and check range than iterate range and then have to check range again.

File: code/datums/gamemodes/gangwar.dm [hunk 2]
```diff
@@ -746,104 +1111,116 @@ proc/broadcast_to_all_gangs(var/message)
...
+			otherTag.owners.unclaim_tiles(target_turf,GANG_TAG_INFLUENCE, GANG_TAG_SIGHT_RANGE)
+			otherTag.disable()
 
 		src.gang.make_tag(target_turf)
-		boutput(M, SPAN_NOTICE("You have claimed this area for your gang!"))
+		S.empty = TRUE
+		S.icon_state = "spraycan_crushed"
+		var/mob/M = owner
+		gang.add_points(round(100), M, showText = TRUE)
+		boutput(M, SPAN_NOTICE("You have claimed this area for your gang and gained bonus points!"))
```
Comment (TobleroneSwordfish) [inline code/datums/gamemodes/gangwar.dm:1156]: this should be logged

File: code/datums/gamemodes/gangwar.dm [hunk 3]
```diff
@@ -746,104 +1111,116 @@ proc/broadcast_to_all_gangs(var/message)
...
-	var/max_health = 200
-	var/health = 200
-	var/damage_warning_timeout = 0
-	var/broken = 0
+	/// the overlay this locker should show, after doing stuff like blinking red for errors
 	var/image/default_screen_overlay = null
 	var/HTML = null
-	var/list/buyable_items = list()
 
+	/// the overlay this locker should show, after doing stuff like blinking red for errors
```
Comment (TobleroneSwordfish) [inline code/datums/gamemodes/gangwar.dm:1171]: Wrong comment

**Review by TobleroneSwordfish — CHANGES_REQUESTED** [review]
Got about halfway through, code looks generally good so far.

**Comment by TDHooligan** [comment]
> Some initial jank in no particular order:

> * Pill bottles should be able to go in gang lockers.
sure. can do

> * The message for pouring a beaker of drugs into a gang locker claims you put the beaker into the locker.
oops

> * The gang locker is non-dense. I think we talked about this at some point, but discord won't let me search threads properly so uhh why is it non-dense?
we had this discussion around the merge. mostly because it's unbreakable and 99% of the time, in tiny areas. if it being a solid locker is a problem it'd just be better as a different sprite/object entirely

> * The round seems to end instantly when the shuttle is auto-called?
i thought this was because of running local for some reason, and never questioned it :shelterwot:

> * The little points popup doesn't display when putting drugs into a locker.
odd. ill investigate

File: code/datums/gamemodes/gangwar.dm [hunk 4]
```diff
@@ -347,7 +345,8 @@ proc/broadcast_to_all_gangs(var/message)
 	var/static/list/uniform_list
 	/// Mask or hat items that are being used by other gangs as part of their gang uniform.
 	var/static/list/headwear_list
-
+	var/static/list/color_list = list("#88CCEE","#117733","#332288","#DDCC77","#CC6677","#AA4499") //(hopefully) colorblind friendly palette
+	var/static/list/colors_left =  list(1,2,3,4,5,6)
```
[the remark this answers is not in the document]
Comment (TDHooligan) [reply code/datums/gamemodes/gangwar.dm:349]: it is pretty cursed. the issue with this is that it's used to track which colors are used, but also uses the static color_list entry for the gang icons (ie. gang_head_3 for 332288).

[the remark this answers is not in the document]
Comment (TDHooligan) [reply code/datums/gamemodes/gangwar.dm:349]: thoughts on a saner fix?

File: code/datums/gamemodes/gangwar.dm [hunk 5]
```diff
@@ -629,6 +813,137 @@ proc/broadcast_to_all_gangs(var/message)
...
+		else if (length(mindList) == 0) //we have no choice but to pick amongst the blacklist.
+			return get_random_civvie()
+		else
+			return pick(mindList)
+
+	/// collects and remembers all valid areas to spawn loot/crates
+	proc/find_potential_drop_zones()
+		potential_drop_zones = list()
+		var/list/area/areas = get_accessible_station_areas()
+		for(var/k in areas)
```
[the remark this answers is not in the document]
Comment (TDHooligan) [reply code/datums/gamemodes/gangwar.dm:848]: almost sorry :)

File: _std/types.dm [hunk 6]
```diff
@@ -217,6 +217,7 @@ var/list/list/by_cat = list()
 #define TR_CAT_AREA_PROCESS "process_area"
 #define TR_CAT_RANCID_STUFF "rancid_stuff"
 #define TR_CAT_GHOST_OBSERVABLES "ghost_observables"
+#define TR_CAT_GANGTAGS "gang_tags"
```
[the remark this answers is not in the document]
Comment (TDHooligan) [reply _std/types.dm:220]: yeah the tracking cat died a while ago as i worked through different ways to handle tags. good spot

File: code/datums/gamemodes/gangwar.dm [hunk 7]
```diff
@@ -39,6 +30,8 @@
 	var/shuttle_called = FALSE
 	var/mob/kidnapping_target
 
+	var/gangtag_scheduler = new /datum/controller/processScheduler()
```
[the remark this answers is not in the document]
Comment (TDHooligan) [reply code/datums/gamemodes/gangwar.dm:33]: oh fuck. thats not good

File: code/datums/gamemodes/gangwar.dm [hunk 8]
```diff
@@ -1640,18 +2364,99 @@ proc/broadcast_to_all_gangs(var/message)
...
+	icon = 'icons/obj/decals/graffiti.dmi'
+	icon_state = "gangtag0"
+	var/datum/gang/owners = null
+	var/list/mobs
+	var/heat = 0 // a rough estimation of how regularly this tag has people near it
+	var/image/heatTracker
+	var/active = TRUE
+	/// Deletes all duplicate tags (IE, from the same gang) on this tile
+	proc/delete_same_tags()
+		for(var/obj/decal/gangtag/T in get_turf(src))
```
Comment (TobleroneSwordfish) [inline code/datums/gamemodes/gangwar.dm:2384]: T means turf, don't use single letter variable names for things that aren't in ATOM generally.

File: code/datums/gamemodes/gangwar.dm [hunk 9]
```diff
@@ -1640,18 +2364,99 @@ proc/broadcast_to_all_gangs(var/message)
...
+			if(T.owners == src.owners && T != src) qdel(T)
+
+	/// Makes this tag insert, so it no longer provides points.
```
Comment (TobleroneSwordfish) [inline code/datums/gamemodes/gangwar.dm:2387]:
```suggestion
	/// Makes this tag inert, so it no longer provides points.
```

File: code/datums/gamemodes/gangwar.dm [hunk 10]
```diff
@@ -1640,18 +2364,99 @@ proc/broadcast_to_all_gangs(var/message)
...
+			else
+				mappedHeat = round(max(0,calculatedHeat))+1
+
+
+		var/score = 0
+		score = ceil(mappedHeat * GANG_TAG_POINTS_PER_HEAT)
+		owners.score_turf += score
+		owners.add_points(score)
+		owners.show_score_maptext(score, get_turf(src))
+		heatTracker.icon_state = "gang_heat_[mappedHeat]"
```
Comment (TobleroneSwordfish) [inline code/datums/gamemodes/gangwar.dm:2427]: This should probably have some comments to explain what all the different heat vars mean

File: code/datums/gamemodes/gangwar.dm [hunk 11]
```diff
@@ -1640,18 +2364,99 @@ proc/broadcast_to_all_gangs(var/message)
...
+
+	New()
+		..()
+		START_TRACKING
+		for(var/obj/decal/gangtag/T in get_turf(src))
```
Comment (TobleroneSwordfish) [inline code/datums/gamemodes/gangwar.dm:2432]: Same here, single letter vars bad

File: code/datums/jobs.dm [hunk 12]
```diff
@@ -3055,6 +3055,28 @@ ABSTRACT_TYPE(/datum/job/special/pod_wars)
 	limit = -1
 	wiki_link = "https://wiki.ss13.co/Game_Modes#Football"
 
+
+/datum/job/special/gang_respawn
```
Comment (TobleroneSwordfish) [inline code/datums/jobs.dm:3059]: Just to check, this doesn't show up on the join announcements as "Person has signed up as Gang Respawn"?

File: code/datums/mind.dm [hunk 13]
```diff
@@ -29,6 +29,10 @@ datum/mind
 	/// A list of every antagonist datum subordinate to this mind.
 	var/list/datum/antagonist/subordinate/subordinate_antagonists = list()
 
+	//Gang variables
+	var/obj/item/device/pda2/originalPDA //! The PDA that this crewmember started with - for gang PDA messages
+	var/gang_points = 0 //! The amount of points this mind has to redeem in lockers.
```
Comment (TobleroneSwordfish) [inline code/datums/mind.dm:34]: Not great for this to be on mind, could just be an associative list on the gang datum? Or on the antagonist datum maybe

File: code/modules/admin/debug.dm [hunk 14]
```diff
@@ -711,7 +711,7 @@ body
 	boutput(usr, "Gang scores:")
 
 	for(var/datum/gang/G in get_all_gangs())
-		boutput(usr, "[G.gang_name]: [G.gang_score()] ([G.num_areas_controlled()] areas)")
+		boutput(usr, "[G.gang_name]: [G.gang_score()] ([G.num_tiles_controlled()] areas)")
```
Comment (TobleroneSwordfish) [inline code/modules/admin/debug.dm:714]:
```suggestion
		boutput(usr, "[G.gang_name]: [G.gang_score()] ([G.num_tiles_controlled()] tiles)")
```

File: code/modules/antagonists/gang/abilities/_gang_leader_ability_holder.dm [hunk 15]
```diff
@@ -138,6 +137,67 @@
...
+
+		if (!M.mind || !M.get_gang())
+			boutput(M, SPAN_ALERT("Gang lockers? Huh?"))
+			return TRUE
+		var/datum/gang/userGang = M.get_gang()
+		var/obj/ganglocker/locker = userGang.locker
+		if (!locker)
+			boutput(M, SPAN_ALERT("Your gang doesn't have a locker!"))
+			return TRUE
+		if (M.GetComponent(/datum/component/tracker_hud))
```
Comment (TobleroneSwordfish) [inline code/modules/antagonists/gang/abilities/_gang_leader_ability_holder.dm:163]:
```suggestion
		if (M.GetComponent(/datum/component/tracker_hud/gang))
```

File: code/modules/antagonists/gang/abilities/_gang_leader_ability_holder.dm [hunk 16]
```diff
@@ -165,27 +221,46 @@
...
+			return
+
+		area = get_area(M)
+
+		if((area.teleport_blocked) || istype(area, /area/supply) || istype(area, /area/shuttle/))
+			boutput(M, SPAN_ALERT("You can't place your base here!"))
+			return
+		if (length(T.controlling_gangs))
+			boutput(M, SPAN_ALERT("You can't place your base in another gang's turf!"))
+			return
```
Comment (TobleroneSwordfish) [inline code/modules/antagonists/gang/abilities/_gang_leader_ability_holder.dm:252]: Copy-paste, make this a proc

File: code/modules/antagonists/gang/gang_member.dm [hunk 17]
```diff
@@ -27,6 +31,16 @@
 		if (!ishuman(src.owner.current))
 			return FALSE
 
+
+		var/datum/abilityHolder/gang/A = src.owner.current.get_ability_holder(/datum/abilityHolder/gang)
```
Comment (TobleroneSwordfish) [inline code/modules/antagonists/gang/gang_member.dm:35]: A

File: code/modules/status_system/statusEffects.dm [hunk 18]
```diff
@@ -1279,17 +1279,20 @@
 		var/const/max_health = 30
 		var/const/max_stam = 60
 		var/const/regen_stam = 5
-		var/const/max_dist = 50
 		var/mob/living/carbon/human/H
 		var/datum/gang/gang
 		var/on_turf = 0
 
+		var/elapsedTime = 0
```
Comment (TobleroneSwordfish) [inline code/modules/status_system/statusEffects.dm:1286]: Could just use `ON_COOLDOWN` instead

File: code/modules/status_system/statusEffects.dm [hunk 19]
```diff
@@ -1308,37 +1311,80 @@
...
+			var/mob/living/carbon/human/H
+			if(!ishuman(owner))
+				return
+			H = owner
+			elapsedTime += timePassed
+			var/times = (elapsedTime / tickSpacing)
+			if(times >= 1)
+				elapsedTime -= (round(times) * tickSpacing)
+				for(var/i in 1 to times)
+					H.HealDamage("All", 0.2, 0.2, 0)
```
Comment (TobleroneSwordfish) [inline code/modules/status_system/statusEffects.dm:1323]: Not a huge fan of magical passive healing, maybe only when on low health/in crit?

File: code/obj/item/device/pda2/pda2.dm [hunk 20]
```diff
@@ -378,6 +378,9 @@
 		if(ismob(src.loc))
 			var/mob/mob = src.loc
 			get_all_character_setup_ringtones()
+			if(mob.mind)
```
Comment (TobleroneSwordfish) [inline code/obj/item/device/pda2/pda2.dm:381]: Questionable place to put this, might get overridden in some odd edge cases. Maybe better in the job setup/character spawn code?

**Review by TobleroneSwordfish — CHANGES_REQUESTED** [review]
the rest!

File: code/datums/gamemodes/gangwar.dm [hunk 21]
```diff
@@ -935,6 +1337,110 @@ proc/broadcast_to_all_gangs(var/message)
...
+		boutput(user, "Hunting for a new member...")
+		try_gang_respawn(user)
+
+	/// Respawns a mind as a new gang member
+	proc/gang_respawn(var/datum/mind/target)
+		if (target.get_antagonist("gang_leader"))
+			target.remove_antagonist(ROLE_GANG_LEADER)
+		else
+			target.remove_antagonist(ROLE_GANG_MEMBER)
+		var/mob/living/carbon/human/normal/P = new/mob/living/carbon/human/normal(src.loc)
```
[the remark this answers is not in the document]
Comment (TDHooligan) [reply code/datums/gamemodes/gangwar.dm:1386]: P!!!!!!!!

File: code/datums/gamemodes/gangwar.dm (hunk 1 above)
Comment (TDHooligan) [reply code/datums/gamemodes/gangwar.dm:957]: i forgor how to do this

File: code/datums/mind.dm (hunk 13 above)
Comment (TDHooligan) [reply code/datums/mind.dm:34]: yea could work

File: code/datums/jobs.dm (hunk 12 above)
Comment (TDHooligan) [reply code/datums/jobs.dm:3059]: mhm

File: code/modules/status_system/statusEffects.dm (hunk 19 above)
Comment (TDHooligan) [reply code/modules/status_system/statusEffects.dm:1323]: idk. admittedly it existed in the old version and i facelifted it. for now it further encourages gang members stickin near their locker?

File: code/modules/status_system/statusEffects.dm (hunk 18 above)
Comment (TDHooligan) [reply code/modules/status_system/statusEffects.dm:1286]: this accounts for mult atm. is there an alternative that does this?

File: code/obj/item/device/pda2/pda2.dm (hunk 20 above)
Comment (TDHooligan) [reply code/obj/item/device/pda2/pda2.dm:381]: don't disagree. also really, really dont have a clue here though

File: code/datums/gamemodes/gangwar.dm [hunk 22]
```diff
@@ -935,6 +1337,110 @@ proc/broadcast_to_all_gangs(var/message)
...
+			return
+
+		if (length(src.gang.members) >= GANG_MAX_MEMBERS)
+			if (src.gang.living_member_count() >= GANG_MAX_MEMBERS)
+				boutput(user, "You've got a full gang!")
+				return
+			else
+				boutput(user, "You've got a full gang! Choose a dead member to hire over.")
+				var/list/datum/mind/members = gang.get_dead_memberlist()
+				var/datum/mind/chosenPlayer = input("Select a gang member to remove.", "Remove gang member") as null|anything in members
```
[the remark this answers is not in the document]
Comment (TDHooligan) [reply code/datums/gamemodes/gangwar.dm:1381]: dead_memberlist returns an associative list of [current real_name] -> mind, which is how they stringify in the list

File: code/modules/status_system/statusEffects.dm (hunk 18 above)
Comment (TobleroneSwordfish) [reply code/modules/status_system/statusEffects.dm:1286]: `ON_COOLDOWN` uses `timeofday` so should be adjusted anyway, it's not dependent on server tick rate (I think)

File: code/datums/gamemodes/gangwar.dm (hunk 1 above)
Comment (TobleroneSwordfish) [reply code/datums/gamemodes/gangwar.dm:967]: `for_by_tcl` or iterate over `by_type`, pretty sure you've done it somewhere else in here

File: code/obj/item/device/pda2/pda2.dm (hunk 20 above)
Comment (TobleroneSwordfish) [reply code/obj/item/device/pda2/pda2.dm:381]: `spawnId` looks like a good proc? That's the code that handles renaming PDAs so makes some sense.

**Review by TobleroneSwordfish — CHANGES_REQUESTED** [review]
Actually it looks like gang tags might not be working right? They seem to always be at max heat, possibly because the var they use to track mobs is called `mobs` which shadows the global list `mobs` and you're accessing it without using `src`

**Comment by TDHooligan** [comment]
> Actually it looks like gang tags might not be working right? They seem to always be at max heat, possibly because the var they use to track mobs is called `mobs` which shadows the global list `mobs` and you're accessing it without using `src`

they seem ok to me. they're all relative though so you need at least 2 tags to compare (the hottest tag is what 'max heat' is)

File: code/datums/gamemodes/gangwar.dm [hunk 23]
```diff
@@ -1040,56 +1579,180 @@ proc/broadcast_to_all_gangs(var/message)
...
 				boutput(user, SPAN_ALERT("<b>You cant stash toy guns in the locker</b>"))
 
-				return 0
-			// var/obj/item/gun/gun = item
-			gang.score_gun += round(300)
-			gang.spendable_points += round(300)
+				return
 
+			if(istype(item, /obj/item/gun/kinetic/slamgun))
+				boutput(user, SPAN_ALERT("<b>You cant stash toy guns in the locker</b>"))
```
Comment (TobleroneSwordfish) [inline code/datums/gamemodes/gangwar.dm:1609]:
```suggestion
				boutput(user, SPAN_ALERT("<b>This shoddy firearm is worth a lot less</b>"))
```
Also maybe include zipguns if you want to make makeshift stuff worth less?

**Review by TobleroneSwordfish — CHANGES_REQUESTED** [review]

File: code/datums/gamemodes/gangwar.dm (hunk 23 above)
Comment (TDHooligan) [reply code/datums/gamemodes/gangwar.dm:1609]: nah, zipguns require staplers which are a bit more annoying to get ahold of (and have staffie charm)

File: code/datums/gamemodes/gangwar.dm [hunk 24]
```diff
@@ -1368,6 +2095,9 @@ proc/broadcast_to_all_gangs(var/message)
 		src.gang = gang
 		..()
 
+	random_gangs
```
Comment (TobleroneSwordfish) [inline code/datums/gamemodes/gangwar.dm:2098]: Minor thing: this still says it's full of fliers

**Review by TobleroneSwordfish — APPROVED** [review]

**Pull request merged**